source/PlayerInfo.cpp

```diff
@@ -1136,24 +1145,62 @@ void PlayerInfo::Land(UI *ui)
 	for(const shared_ptr<Ship> &ship : ships)
 		ship->UnloadBays();
 	
-	// Ships that are landed with you on the planet should fully recharge
-	// and pool all their cargo together. Those in remote systems restore
-	// what they can without landing.
+	// Ships that are landed with you on the planet should fully recharge and pool
+	// all their cargo together. Those that are in remote systems or cannot land
+	// anywhere in the player's system restore what they can without landing.
 	bool hasSpaceport = planet->HasSpaceport() && planet->CanUseServices();
 	UpdateCargoCapacities();
 	for(const shared_ptr<Ship> &ship : ships)
 		if(!ship->IsParked() && !ship->IsDisabled())
 		{
 			if(ship->GetSystem() == system)
 			{
-				ship->Recharge(hasSpaceport);
-				ship->Cargo().TransferAll(cargo);
-				if(!ship->GetPlanet())
-					ship->SetPlanet(planet);
+				if(planet->CanLand(*ship) || (ship->IsCarried() && ship->GetParent()->GetSystem() == system
```

Comment:
All carried ships have been unloaded on line 1145 (new numbering), so I expect that they will not have a parent at this point.
Also, carried ships don't have a system while carried if I remember correctly, so if the ship would still have been carried, then the conditions in the if-statement on line 1156 would have returned false for such carried ships.